### PR TITLE
[DO NOT MERGE] Autograd Lab: write custom function and gradcheck

### PR DESCRIPTION
**Pull request by saketh-are**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60879
* #60878
* **#60877**

### CODE REVIEW

File: attn_gradcheck.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    @staticmethod
+    def backward(ctx, grad_output, grad_attn):
+        (q, k, v, attn) = ctx.saved_tensors
+
+        # computes and stores sech^2(qk')
+        partial = 1 / torch.cosh(q.matmul(k.swapaxes(0, 1))).pow(2)
+
+        # computes grad_output terms
+        grad_q = grad_output.matmul(v.swapaxes(0, 1)).mul(partial).matmul(k)
```
Comment (albanD) [inline attn_gradcheck.py:21]: nit: you can actually pre-compute `grad_output.matmul(v.swapaxes(0, 1)).mul(partial)` and re-use it here.

File: attn_gradcheck.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import torch
+
+class Attn(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, q, k, v):
+        attn = torch.matmul(q, k.swapaxes(0, 1))
```
Comment (albanD) [inline attn_gradcheck.py:6]: note that in general, we would add error checking here to make sure that we don't get un-expected size.

**Review by albanD — COMMENTED** [review]
Looks good!

File: attn_gradcheck.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        attn = torch.tanh(attn)
+        output = torch.matmul(attn, v)
+
+        ctx.save_for_backward(q, k, v, attn)
+        return output, attn
+
+    @staticmethod
+    def backward(ctx, grad_output, grad_attn):
```
Comment (albanD) [inline attn_gradcheck.py:14]: Note that you could be using `needs_input_grad` here to make the backward faster in the case where not all inputs to the forward have requires_grad = True

File: attn_gradcheck.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import torch
+
+class Attn(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, q, k, v):
+        assert q.size() == k.size()
+        assert q.size()[0] == v.size()[0]
```
Comment (zou3519) [inline attn_gradcheck.py:7]: Question [on the lab](https://github.com/pytorch/pytorch/wiki/Autograd-Onboarding-Lab) actually: are q, k, v defined to be 2D tensors? If so then these should be checks that they're all two dimensional

**Pull request closed without merging**